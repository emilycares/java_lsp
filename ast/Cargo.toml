[package]
name = "ast"
version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
documentation.workspace = true
readme.workspace = true
license.workspace = true

[dependencies]
phf.workspace = true
smol_str.workspace = true

[dev-dependencies]
insta.workspace = true
criterion.workspace = true

[[bench]]
name = "base"
harness = false
