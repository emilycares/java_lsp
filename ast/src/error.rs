use std::panic::Location;

use crate::skip_helper::skip;

use super::lexer::{PositionToken, Token};

pub trait PrintErr {
    fn print_err(&self, content: &str);
}

impl<T> PrintErr for Result<T, AstError> {
    fn print_err(&self, content: &str) {
        match self {
            Ok(_) => (),
            Err(e) => e.print_err(content),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum AstError {
    ExpectedToken(ExpectedToken),
    InvalidAvailability(InvalidToken),
    InvalidJtype(InvalidToken),
    UnexpectedEOF(String, u32, u32),
    IdentifierEmpty(InvalidToken),
}
impl AstError {
    #[track_caller]
    pub fn eof() -> Self {
        let loc = Location::caller();
        Self::UnexpectedEOF(loc.file().to_string(), loc.line(), loc.column())
    }
    pub fn print_err(&self, content: &str) {
        match self {
            AstError::ExpectedToken(expected_token) => {
                print_helper(
                    content,
                    expected_token.line,
                    expected_token.col,
                    format!(
                        "Expected token {:?} found: {:?}",
                        expected_token.expected, expected_token.found
                    ),
                );
            }
            AstError::InvalidAvailability(invalid_token) => {
                print_helper(
                    content,
                    invalid_token.line,
                    invalid_token.col,
                    format!(
                        "Invalid Availability token found: {:?} valid onese ar public, private, protected",
                        invalid_token.found
                    ),
                );
            }
            AstError::UnexpectedEOF(file, line, col) => {
                eprintln!("Unexpected end of File: {file}:{line}:{col}")
            }
            AstError::InvalidJtype(invalid_token) => {
                print_helper(
                    content,
                    invalid_token.line,
                    invalid_token.col,
                    format!(
                        "Invalid Type token found: {:?} valid onese ar Int, String",
                        invalid_token.found
                    ),
                );
            }
            AstError::IdentifierEmpty(invalid_token) => {
                print_helper(
                    content,
                    invalid_token.line,
                    invalid_token.col,
                    format!("Identifier empty found: {:?}", invalid_token.found),
                );
            }
        }
    }
}

fn print_helper(content: &str, line: usize, col: usize, msg: String) {
    let is_zero = line == 0;
    let mut lines = match is_zero {
        false => content.lines().enumerate().skip(line - 1),
        true => content.lines().enumerate().skip(line),
    };
    if !is_zero {
        if let Some((number, line)) = lines.next() {
            eprintln!("{number} {line}");
        }
    }
    if let Some((number, line)) = lines.next() {
        eprintln!("{number} \x1b[93m{line}\x1b[0m");
    }
    let spaces = " ".repeat(col);
    eprintln!("  {spaces}^");
    eprintln!("  | {msg}");
    if let Some((number, line)) = lines.next() {
        eprintln!("{number}{line}");
    }
}

pub fn assert_token(
    tokens: &[PositionToken],
    pos: usize,
    expected: Token,
) -> Result<usize, AstError> {
    let pos = skip(tokens, pos)?;
    match tokens.get(pos) {
        Some(t) => {
            if t.token != expected {
                return Err(AstError::ExpectedToken(ExpectedToken::from(t, expected)));
            }
            return Ok(pos + 1);
        }
        None => Err(AstError::eof()),
    }?
}

#[derive(Debug, PartialEq)]
pub struct ExpectedToken {
    pub expected: Token,
    pub found: Token,
    pub line: usize,
    pub col: usize,
}

impl ExpectedToken {
    pub fn from(pos: &PositionToken, expected: Token) -> Self {
        Self {
            expected,
            found: pos.token.clone(),
            line: pos.line,
            col: pos.col,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct InvalidToken {
    pub found: Token,
    pub line: usize,
    pub col: usize,
}

impl InvalidToken {
    pub fn from(pos: &PositionToken) -> Self {
        Self {
            found: pos.token.clone(),
            line: pos.line,
            col: pos.col,
        }
    }
}
