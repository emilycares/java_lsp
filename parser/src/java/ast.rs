use super::{
    error::{AstError, ExpectedToken, InvalidToken, assert_token},
    lexer::{PositionToken, Token},
};

#[derive(Debug, PartialEq)]
pub struct AstFile {
    package: String,
    imports: Vec<String>,
    thing: AstThing,
}
#[derive(Debug, PartialEq)]
pub enum AstAvailability {
    Public,
    Private,
    Protected,
}

#[derive(Debug, PartialEq)]
pub struct AstClass {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub struct AstInterface {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub enum AstThing {
    Class(AstClass),
    Interface(AstInterface),
    None,
}

pub fn parse_file(tokens: &[PositionToken], pos: usize) -> Result<AstFile, AstError> {
    let (package_name, pos) = parse_package(tokens, pos)?;
    let (imports, pos) = parse_imports(tokens, pos)?;
    let (thing, _pos) = parse_thing(tokens, pos)?;

    Ok(AstFile {
        package: package_name,
        imports,
        thing,
    })
}

/// package ch.emilycares;
fn parse_package(tokens: &[PositionToken], pos: usize) -> Result<(String, usize), AstError> {
    let mut next_pos = pos;
    let mut iter = tokens.iter();
    assert_token(&mut iter, Token::Package)?;
    let mut package_name = String::new();
    for t in iter {
        next_pos += 1;
        match &t.token {
            Token::Identifier(id) => package_name.push_str(id),
            Token::Dot => package_name.push('.'),
            Token::Semicolon => break,
            _ => continue,
        }
    }
    Ok((package_name, next_pos))
}

///  import java.io.IOException;
///  import java.net.Socket;
fn parse_imports(tokens: &[PositionToken], pos: usize) -> Result<(Vec<String>, usize), AstError> {
    let mut next_pos = pos + 1;
    let mut imports = vec![];

    while let Ok((import, new_pos)) = parse_import(tokens, next_pos) {
        next_pos = new_pos;
        imports.push(import);
    }

    Ok((imports, next_pos))
}

///  import java.io.IOException;
fn parse_import(tokens: &[PositionToken], pos: usize) -> Result<(String, usize), AstError> {
    let mut iter = tokens.iter().skip(pos);
    assert_token(&mut iter, Token::Import)?;

    let mut pos = pos;
    let mut import_name = String::new();
    for t in iter {
        pos += 1;
        match &t.token {
            Token::Identifier(id) => import_name.push_str(id),
            Token::Dot => import_name.push('.'),
            Token::Star => import_name.push('*'),
            Token::Semicolon => {
                pos += 1;
                break;
            }
            _ => continue,
        }
    }
    Ok((import_name, pos))
}

///  public class Everything { ...
///  public interface Constants { ...
fn parse_thing(tokens: &[PositionToken], pos: usize) -> Result<(AstThing, usize), AstError> {
    let mut pos = pos;
    let mut iter = tokens.iter().skip(pos);
    pos += 1;
    let avaliability = parse_avaliability(iter.next())?;
    match iter.next() {
        Some(t) => match t {
            PositionToken {
                token: Token::Class,
                line: _,
                col: _,
            } => Ok((AstThing::Class(AstClass { avaliability }), pos)),
            PositionToken {
                token: Token::Interface,
                line: _,
                col: _,
            } => Ok((AstThing::Interface(AstInterface { avaliability }), pos)),
            found => Err(AstError::ExpectedToken(ExpectedToken::from(
                found,
                Token::Class,
            ))),
        },
        None => Err(AstError::UnexpectedEOF),
    }
}

fn parse_avaliability(token: Option<&PositionToken>) -> Result<AstAvailability, AstError> {
    let Some(token) = token else {
        return Err(AstError::UnexpectedEOF);
    };
    match token {
        PositionToken {
            token: Token::Public,
            line: _,
            col: _,
        } => Ok(AstAvailability::Public),
        PositionToken {
            token: Token::Private,
            line: _,
            col: _,
        } => Ok(AstAvailability::Private),
        PositionToken {
            token: Token::Protected,
            line: _,
            col: _,
        } => Ok(AstAvailability::Protected),
        e => Err(AstError::InvalidAvailability(InvalidToken::from(e))),
    }
}

#[cfg(test)]
pub mod tests {
    use pretty_assertions::assert_eq;

    use crate::java::error::PrintErr;
    use crate::java::{
        ast::{AstAvailability, AstClass, AstFile, AstInterface, AstThing, parse_file},
        lexer,
    };

    #[test]
    fn everything() {
        let content = include_str!("../../test/Everything.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec![],
                thing: AstThing::Class(AstClass {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }

    #[test]
    fn superee() {
        let content = include_str!("../../test/Super.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec!["java.io.IOException".to_string()],
                thing: AstThing::Class(AstClass {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }

    #[test]
    fn constants() {
        let content = include_str!("../../test/Constants.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec![
                    "jdk.net.Sockets".to_string(),
                    "java.io.IOException".to_string(),
                    "java.net.Socket".to_string()
                ],
                thing: AstThing::Interface(AstInterface {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }
}
