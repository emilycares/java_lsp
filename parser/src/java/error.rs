use super::lexer::{PositionToken, Token};

pub trait PrintErr {
    fn print_err(&self, content: &str);
}

impl<T> PrintErr for Result<T, AstError> {
    fn print_err(&self, content: &str) {
        match self {
            Ok(_) => (),
            Err(e) => e.print_err(content),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum AstError {
    ExpectedToken(ExpectedToken),
    InvalidAvailability(InvalidToken),
    UnexpectedEOF,
}
impl AstError {
    pub fn print_err(&self, content: &str) {
        match self {
            AstError::ExpectedToken(expected_token) => {
                print_helper(
                    content,
                    expected_token.line,
                    expected_token.col,
                    format!(
                        "Expected token {:?} found: {:?}",
                        expected_token.expected, expected_token.found
                    ),
                );
            }
            AstError::InvalidAvailability(invalid_token) => {
                print_helper(
                    content,
                    invalid_token.line,
                    invalid_token.col,
                    format!(
                        "Invalid Availability token found: {:?} valid onese ar public, private, protected",
                        invalid_token.found
                    ),
                );
            }
            AstError::UnexpectedEOF => eprintln!("Unexpected end of File"),
        }
    }
}

fn print_helper(content: &str, line: usize, col: usize, msg: String) {
    let mut lines = content.lines().enumerate().skip(line - 1);
    let (number, line) = lines.next().unwrap();
    println!("{number} {line}");
    let (number, line) = lines.next().unwrap();
    println!("{number} \x1b[93m{line}\x1b[0m");
    let spaces = " ".repeat(col);
    println!("  {spaces}^");
    println!("  | {msg}");
    let (number, line) = lines.next().unwrap();
    println!("{number}{line}");
}

pub fn assert_token<'a>(
    mut iter: impl Iterator<Item = &'a PositionToken>,
    expected: Token,
) -> Result<(), AstError> {
    match iter.next() {
        Some(t) => {
            if t.token != expected {
                return Err(AstError::ExpectedToken(ExpectedToken::from(t, expected)));
            }
            return Ok(());
        }
        None => Err(AstError::UnexpectedEOF),
    }?
}

#[derive(Debug, PartialEq)]
pub struct ExpectedToken {
    pub expected: Token,
    pub found: Token,
    pub line: usize,
    pub col: usize,
}

impl ExpectedToken {
    pub fn from(pos: &PositionToken, expected: Token) -> Self {
        Self {
            expected,
            found: pos.token.clone(),
            line: pos.line,
            col: pos.col,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct InvalidToken {
    pub found: Token,
    pub line: usize,
    pub col: usize,
}

impl InvalidToken {
    pub fn from(pos: &PositionToken) -> Self {
        Self {
            found: pos.token.clone(),
            line: pos.line,
            col: pos.col,
        }
    }
}
