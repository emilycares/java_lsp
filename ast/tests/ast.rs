use ast::error::PrintErr;
use ast::{
    lexer, parse_block_variable, parse_file, parse_lambda, parse_name, parse_recursive_expression,
    parse_string_literal,
};

#[test]
fn everything() {
    let content = include_str!("../../parser/test/Everything.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn skip_comments() {
    let content = include_str!("../test/FullOffComments.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn locale_variable_table() {
    let content = include_str!("../../parser/test/LocalVariableTable.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn superee() {
    let content = include_str!("../../parser/test/Super.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn constants() {
    let content = include_str!("../../parser/test/Constants.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn super_interface() {
    let content = include_str!("../../parser/test/SuperInterface.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn interface_base() {
    let content = include_str!("../../parser/test/InterfaceBase.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn variants() {
    let content = include_str!("../../parser/test/Variants.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn types() {
    let content = include_str!("../../parser/test/Types.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn annotated() {
    let content = include_str!("../../parser/test/Annotated.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn expression_base() {
    let content = "Logger.getLogger(Test.class)";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_recursive_expression(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn expression_array_access() {
    let content = "numbers[0]";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_recursive_expression(&tokens, 0);
    parsed.print_err(content);
    let ast = parsed.unwrap();
    insta::assert_debug_snapshot!(ast);
    assert_eq!(ast.1, tokens.len());
}
#[test]
fn expression_multi_array_access() {
    let content = "numbers[0][0][0]";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_recursive_expression(&tokens, 0);
    parsed.print_err(content);
    let ast = parsed.unwrap();
    insta::assert_debug_snapshot!(ast);
    assert_eq!(ast.1, tokens.len());
}

#[test]
fn equasion_method_call() {
    let content = r#""z" + a.getThing()"#;
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_recursive_expression(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn annotation() {
    let content = include_str!("../../parser/test/Annotation.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn more_syntax() {
    let content = include_str!("../../parser/test/Syntax.java");
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_file(&tokens);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn variable_array() {
    let content = r#"String[] cars = {"Volvo", "BMW", "Ford", "Mazda"};"#;
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_block_variable(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn multi_line_string() {
    let content = r#"
        """
        Here is a muilti
        line
        string""
        "#;
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_string_literal(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn name() {
    let content = "Logger3m3m3m3";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_name(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn lambda() {
    let content = "(n) -> { System.out.println(n); }";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_lambda(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn lambda_1() {
    let content = "n -> { }";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_lambda(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
#[test]
fn lambda_2() {
    let content = "(a, b, c) -> { }";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_lambda(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}

#[test]
fn equal_expr() {
    let content = "a == b ";
    let tokens = lexer::lex(content).unwrap();
    let parsed = parse_recursive_expression(&tokens, 0);
    parsed.print_err(content);
    insta::assert_debug_snapshot!(parsed.unwrap());
}
