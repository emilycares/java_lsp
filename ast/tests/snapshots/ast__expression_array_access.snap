---
source: ast/tests/ast.rs
expression: ast
---
(
    AstRecursiveExpression {
        range: AstRange {
            start: AstPoint {
                line: 0,
                col: 0,
            },
            end: AstPoint {
                line: 0,
                col: 10,
            },
        },
        ident: Some(
            Identifier(
                AstIdentifier {
                    range: AstRange {
                        start: AstPoint {
                            line: 0,
                            col: 0,
                        },
                        end: AstPoint {
                            line: 0,
                            col: 7,
                        },
                    },
                    value: "numbers",
                },
            ),
        ),
        values: None,
        next: Some(
            AstRecursiveExpression {
                range: AstRange {
                    start: AstPoint {
                        line: 0,
                        col: 7,
                    },
                    end: AstPoint {
                        line: 0,
                        col: 10,
                    },
                },
                ident: Some(
                    ArrayAccess(
                        Nuget(
                            Int(
                                AstInt {
                                    range: AstRange {
                                        start: AstPoint {
                                            line: 0,
                                            col: 8,
                                        },
                                        end: AstPoint {
                                            line: 0,
                                            col: 9,
                                        },
                                    },
                                    value: 0,
                                },
                            ),
                        ),
                    ),
                ),
                values: None,
                next: None,
                operator: None,
            },
        ),
        operator: None,
    },
    4,
)
